rfc/003-remote-pipelines/README.md

```diff
@@ -0,0 +1,206 @@
+# Remote Pipelines
+
+- **Status**: WIP
+- **Created**: Sep 14, 2022
+- **ETA**: Oct 14, 2022
+- **Authors**:
+  - [Matthias Vallentin](https://github.com/mavam)
+- **Discussion**: [PR #2577](https://github.com/tenzir/vast/pull/2577)
+
+## Overview
+
+This proposal enhances the local pipeline execution with a remote component,
+making it possible to connect pipeline across process boundaries.
+
+## Problem Statement
+
+As of RFC-001, VAST pipelines can execute locally. But a one-shot pipeline
+execution is not sufficient. After this building block, natural questions arise:
+
+- How can I connect a local and remote pipeline?
+- How can I manage (add/update/remove) a set of continuously running pipelines?
+- How do pipelines interact with VAST server nodes? Specifically, how can I send
+  data into a VAST node, and how can I query data from a VAST node?
+
+This proposal seeks an answer to these questions.
+
+## Solution Proposal
```

Comment:
I think we should create 2 distinct named entities:
- "named pipelines": a chain of operators that will execute provided that data will flow into its first operator
- "named dataflows" (not sure about the term, but we can use it here as it isn't too overloaded yet): a specific point in the graph of data flows to which you can attach a pipelines input or ouput

In the plumbing analogy that we, Mario loving data engineers, love: "pipelines" are pipes, "dataflows" are logical sources/sinks

This decoupling has many benefits:
- you can combine the output of M pipelines and subscribe to that with N other pipelines
- these named dataflows are very similar to the "topic" concept in the Fabric/Threatbus, it will provide good abstraction points for domain specific abstract "streams of interest". For instance, you could have a default `ocsf_finding` "named dataflow". Every app wanting to contribute to it can create a pipeline that pushes to it and every app wanting to subscribe to all alerts will now that this is the best place to start.

The example bellow would look something like this:
```bash
vast spawn pipe1 'from - | read zeek | where orig_bytes > 1 GiB | push zeek_large'
vast exec 'pull zeek_large | where resp_bytes > 1 KiB'
```

Reply:
I like the additional concept of "dataflow" (or just "flow") as abstraction for pipeline composition.

You could argue that once you have a non-void operator, you can turn that into a flow by giving it a name, and optionally binding it to remote endpoint. 

For example:

```
load zeek://1.2.3.4 | where #type == "zeek.notice"
```

This pipeline has type `<Void, Arrow>`. In step one, you would give it a name:

```
vast spawn zeek-alerts 'load zeek://1.2.3.4 | where #type == "zeek.notice"'
```

If you want to use it outside of your current VAST node, you would have to expose it in some fashion, e.g., to a service registry:

```
vast publish zeek-alerts --endpoint 1.2.3.4:59232
```